### PR TITLE
Feat: Let the ingress-gateway uses the same port as the business service as a proxy for layer 7 traffic.

### PR DESCRIPTION
**Pull request by panniyuyu**
Let the ingress-gateway uses the same port as the business service as a proxy for layer 7 traffic.

### CODE REVIEW

File: plugin/metaprotocol/generator.go [hunk 1]
```diff
@@ -35,6 +37,49 @@ func NewGenerator() *Generator {
 
 // Generate create EnvoyFilters for MetaProtocol services
 func (*Generator) Generate(context *model.EnvoyFilterContext) ([]*model.EnvoyFilterWrapper, error) {
+	if context.Gateway != nil {
+		return generateGatewayEnvoyFilters(context)
+	} else {
```
Comment (Xunzhuo) [inline plugin/metaprotocol/generator.go:42]: else is no need here

File: plugin/metaprotocol/generator.go [hunk 2]
```diff
@@ -35,6 +37,49 @@ func NewGenerator() *Generator {
...
+		envoyfilters = append(envoyfilters,
+			envoyfilter.GenerateReplaceNetworkFilter(
+				context.ServiceEntry,
+				port,
+				outboundProxy,
+				nil,
+				"envoy.filters.network.meta_protocol_proxy",
+				"type.googleapis.com/aeraki.meta_protocol_proxy.v1alpha.MetaProtocolProxy")...)
+		// append workloadSelector for OutboundListener EnvoyFilter
+		for i, _ := range envoyfilters {
```
Comment (Xunzhuo) [inline plugin/metaprotocol/generator.go:70]: To iterate the key, plz just use `for i := range envoyfilters {`

File: test/e2e/metaprotocol-gateway/metaprotocol_test.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	util.KubeDelete("meta-thrift", "testdata/thrift-sample.yaml", "")
+	util.KubeDelete("istio-system", "../../../k8s/aeraki-bootstrap-config.yaml", "")
+	util.DeleteNamespace("meta-thrift", "")
+}
+
+func TestThriftRouter(t *testing.T) {
+	util.WaitForDeploymentsReady("meta-thrift", 10*time.Minute, "")
+	util.WaitForDeploymentsReady("istio-system", 10*time.Minute, "")
+
+	svcIp, err := util.GetServiceIP("istio-ingressgateway", "istio-system", "")
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/metaprotocol_test.go:60]: `svcIp` -> `svcIP`

File: test/e2e/metaprotocol-gateway/testdata/destinationrule.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  namespace: meta-thrift
+spec:
+  host: thrift-sample-server.meta-thrift.svc.cluster.local
+  subsets:
+    - labels:
+        version: v1
+      name: v1
+    - labels:
+        version: v2
+      name: v2
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/testdata/destinationrule.yaml:14]: A new line is needed at the end of yaml.

File: test/e2e/metaprotocol-gateway/testdata/ingress-gateway.yaml [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+      port:
+        name: tcp-grafana
+        number: 3000
+        protocol: tcp
+    - hosts:
+        - '*'
+      port:
+        name: tcp-metaprotocol-thrift-9090
+        number: 9090
+        protocol: tcp
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/testdata/ingress-gateway.yaml:350]: ditto

File: test/e2e/metaprotocol-gateway/testdata/ingress-patch.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+          name: workload-certs
+      volumes:
+      - configMap:
+          defaultMode: 420
+          name: aeraki-bootstrap-config
+        name: custom-bootstrap-volume
+      - emptyDir:
+        name: workload-socket
+      - emptyDir:
+        name: workload-certs
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/testdata/ingress-patch.yaml:25]: ditto

File: test/e2e/metaprotocol-gateway/testdata/metarouter-v1.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    - match:
+        attributes:
+          method:
+            exact: ""
+      route:
+        - destination:
+            host: thrift-sample-server.meta-thrift.svc.cluster.local
+            port:
+              number: 9090
+            subset: v1
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/testdata/metarouter-v1.yaml:21]: ditto

File: test/e2e/metaprotocol-gateway/testdata/metarouter-v2.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    - match:
+        attributes:
+          method:
+            exact: ""
+      route:
+        - destination:
+            host: thrift-sample-server.meta-thrift.svc.cluster.local
+            port:
+              number: 9090
+            subset: v2
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/testdata/metarouter-v2.yaml:21]: ditto

File: test/e2e/metaprotocol-gateway/testdata/thrift-sample.yaml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+metadata:
+  name: thrift-sample-server
+spec:
+  selector:
+    app: thrift-sample-server
+  ports:
+    - name: tcp-metaprotocol-thrift-9090
+      protocol: TCP
+      port: 9090
+      targetPort: 9090
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/testdata/thrift-sample.yaml:59]: ditto

File: test/e2e/metaprotocol-gateway/testdata/virtualservice.yaml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  - istio-system/istio-ingressgateway
+  hosts:
+  - thrift-sample-server.meta-thrift.svc.cluster.local
+  tcp:
+  - match:
+    - port: 9090
+    route:
+    - destination:
+        host: thrift-sample-server.meta-thrift.svc.cluster.local
+      weight: 100
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/testdata/virtualservice.yaml:17]: ditto

File: test/e2e/metaprotocol-gateway/hello.thrift [hunk 11]
```diff
@@ -0,0 +1,5 @@
+namespace java org.aeraki
+
+service HelloService {
+  string sayHello(1:required string name)
+}
```
Comment (Xunzhuo) [inline test/e2e/metaprotocol-gateway/hello.thrift:5]: ditto

**Review by Xunzhuo — CHANGES_REQUESTED** [review]
Overall looks good to me, some nits. Thanks.

**Comment by panniyuyu** [comment]
OK, i will address these suggestions later. Thanks for reviewing. @Xunzhuo

**Comment by zhaohuabing** [comment]
@panniyuyu Could you please add the gateway e2e test to the github workflow?

**Review by zhaohuabing — CHANGES_REQUESTED** [review]
Could you please add the gateway e2e test to the github workflow?

**Comment by zhaohuabing** [comment]
Thank you, @panniyuyu, for contributing this incredibly useful feature! I just wanted to let you know that since this PR is relatively large, it may take me some time to review and fully understand all of the changes. However, I really appreciate your hard work and am excited to see how this feature can improve our projec

**Comment by panniyuyu** [comment]
> Could you please add the gateway e2e test to the github workflow?

Sure, i will finish it later.

File: test/e2e/metaprotocol-gateway/testdata/ingress-gateway.yaml [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+  name: istio-ingressgateway
+  namespace: istio-system
+spec:
+  selector:
+    app: istio-ingressgateway
+    istio: ingressgateway
+  servers:
+    - hosts:
+        - '*'
+      port:
```
Comment (zhaohuabing) [inline test/e2e/metaprotocol-gateway/testdata/ingress-gateway.yaml:335]: Do we need this port?

File: test/e2e/metaprotocol-gateway/testdata/ingress-gateway.yaml [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+        name: tcp-kaili
+        number: 20001
+        protocol: tcp
+    - hosts:
+        - '*'
+      port:
+        name: tcp-grafana
```
Comment (zhaohuabing) [inline test/e2e/metaprotocol-gateway/testdata/ingress-gateway.yaml:342]: Do we need this port?

**Comment by zhaohuabing** [comment]
@panniyuyu Great job on the PR! I have reviewed it and everything looks good to me. Have you been able to identify the root cause of the failed gateway test? Once you have fixed the test and the yamllint errors, we will be able to merge your changes into the master branch.

File: .github/workflows/e2e-metaprotocol-gateway.yaml [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+      - name: Prepare envrionment
+        run: bash ${SCRIPTS_DIR}/pre.sh
+      - name: Install Minikube
+        run: bash ${SCRIPTS_DIR}/minikube.sh start
+      - name: Install Istio
+        run: bash ${SCRIPTS_DIR}/istio.sh -y -f ${COMMON_DIR}/istio-config.yaml
+      - name: Install aeraki
+        run: bash ${SCRIPTS_DIR}/aeraki.sh
+      - name: test
+        run: go test -v github.com/aeraki-mesh/aeraki/test/e2e/metaprotocol-gateway/ -run TestThriftRouter
```
Comment (Xunzhuo) [inline .github/workflows/e2e-metaprotocol-gateway.yaml:40]: plz remember a new line at yaml files, otherwise CI wont pass. Thanks

**Comment by zhaohuabing** [comment]
> Side comment, we need a document to decribe this new feature, thanks.

The documentation is not a blocker and can be added separately through another PR.

File: test/e2e/metaprotocol-gateway/testdata/virtualservice.yaml
Comment (zhaohuabing) [inline test/e2e/metaprotocol-gateway/testdata/virtualservice.yaml:1]: We don't need this virtualservice, do we?

File: test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+	//shutdown()
+	os.Exit(code)
+}
+func setup() {
+	util.CreateNamespace("meta-thrift", "")
+	util.LabelNamespace("meta-thrift", "istio-injection=disabled", "")
+	util.KubeApply("istio-system", "../../../k8s/aeraki-bootstrap-config.yaml", "")
+	util.KubeApply("meta-thrift", "testdata/thrift-sample.yaml", "")    // thrift server/svc
+	util.KubeApply("istio-system", "testdata/ingress-gateway.yaml", "") // ingress gateway
+	util.KubeApply("meta-thrift", "testdata/virtualservice.yaml", "")
```
Comment (zhaohuabing) [inline test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go:42]: Remove this line

File: test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+	util.KubeApply("meta-thrift", "testdata/destinationrule.yaml", "")
+
+}
+
+func shutdown() {
+	util.KubeDelete("meta-thrift", "testdata/destinationrule.yaml", "")
+	util.KubeDelete("meta-thrift", "testdata/virtualservice.yaml", "")
```
Comment (zhaohuabing) [inline test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go:49]: Remove this line.

File: test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+	code := m.Run()
+	//shutdown()
+	os.Exit(code)
+}
+func setup() {
+	util.CreateNamespace("meta-thrift", "")
+	util.LabelNamespace("meta-thrift", "istio-injection=disabled", "")
+	util.KubeApply("istio-system", "../../../k8s/aeraki-bootstrap-config.yaml", "")
+	util.KubeApply("meta-thrift", "testdata/thrift-sample.yaml", "")    // thrift server/svc
+	util.KubeApply("istio-system", "testdata/ingress-gateway.yaml", "") // ingress gateway
```
Comment (zhaohuabing) [inline test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go:41]: We need to bind a default MetaRouter to the gateway.

File: plugin/metaprotocol/generator.go [hunk 18]
```diff
@@ -35,6 +37,48 @@ func NewGenerator() *Generator {
...
+			envoyfilter.GenerateReplaceNetworkFilter(
+				context.ServiceEntry,
+				port,
+				outboundProxy,
+				nil,
+				"envoy.filters.network.meta_protocol_proxy",
+				"type.googleapis.com/aeraki.meta_protocol_proxy.v1alpha.MetaProtocolProxy")...)
+		// append workloadSelector for OutboundListener EnvoyFilter
+		for i := range envoyfilters {
+			envoyfilters[i].Name = fmt.Sprintf("aeraki-router-outbound--%s-%d", context.Gateway.Name, port.Number)
```
Comment (zhaohuabing) [inline plugin/metaprotocol/generator.go:70]: The current EnvoyFilter name looks like ```aeraki-router-outbound--istio-ingressgateway-909```.  The word router in EnvoyFilter name is too vague. How about  ```aeraki-gateway-$(gateway-name)-$(port)``` ?

File: test/e2e/metaprotocol-gateway/testdata/virtualservice.yaml
Comment (panniyuyu) [reply test/e2e/metaprotocol-gateway/testdata/virtualservice.yaml:1]: We need this virtualservice to let ingressgateway  build listener.

File: test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go (hunk 15 above)
Comment (panniyuyu) [reply test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go]: ditto

File: test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go (hunk 16 above)
Comment (panniyuyu) [reply test/e2e/metaprotocol-gateway/metaprotocol_gateway_test.go]: ditto

File: plugin/metaprotocol/generator.go (hunk 18 above)
Comment (panniyuyu) [reply plugin/metaprotocol/generator.go]: It's reasonable.

**Comment by panniyuyu** [comment]
I have aready fixed the gateway test, yamllint errors. But i have no idea about golang lint errors. Could you please do me a favor? @zhaohuabing

**Comment by zhaohuabing** [comment]
> I have aready fixed the gateway test, yamllint errors. But i have no idea about golang lint errors. Could you please do me a favor? @zhaohuabing 

golang lint checks the order of import packages. This should be fixed with:

goimports -local github.com/aeraki-mesh/aeraki ./...

**Comment by zhaohuabing** [comment]
This PR is the implementation of https://github.com/aeraki-mesh/aeraki/issues/303

**Pull request merged**